File: 0865-smallest-subtree-with-all-the-deepest-nodes/0865-smallest-subtree-with-all-the-deepest-nodes.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def subtreeWithAllDeepest(self, root: TreeNode) -> TreeNode:
        
        def find_node(node, level):
            nonlocal max_level, sub_node
            if(not node):
                return level
            
            left = find_node(node.left, level + 1)
            right = find_node(node.right, level + 1)
            if(left == right and max_level <= left):
                max_level = left
                sub_node = node
            
            return max(left, right)
            
        
        max_level = 0
        sub_node = root
        find_node(root, 1)
        return sub_node
        

File: 0865-smallest-subtree-with-all-the-deepest-nodes/0865-smallest-subtree-with-all-the-deepest-nodes.go
func subtreeWithAllDeepest(root *TreeNode) *TreeNode {
    var node *TreeNode
    _, node = getDeepest(root)
    return node
}

func getDeepest(node *TreeNode) (int32, *TreeNode) {
    if node == nil {
        return 0, nil
    }

    left, leftNode := getDeepest(node.Left)
    right, rightNode := getDeepest(node.Right)

    switch {
    case left == right:
        return left + 1, node
    case left > right:
        return left + 1, leftNode
    default:
        return right + 1, rightNode
    }
}
